sprint_11

Gvozdkov opened the pull request:
Для меня очень тяжелый спринт и возможно, самый сложный)) Я буду надеятся, чти все выполнил правильно ...)

Commit:
Подключена зависимость https://github.com/relatedcode/ProgressHUD и добавлен индикатор загрузки в SplashViewController.swift /  extension SplashViewController: AuthViewControllerDelegate

Commit:
Добавлен новый класс UIBlockingProgressHUD в файл UIBlockingProgressHUD.swift, который синхронизирует показ и скрытие ProgressHUD с блокировкой UI. Сразу после закрытия WebView при авторизации показывает индикатор активности, используя UIBlockingProgressHUD. При завершении запроса, независимо от результата  индикатор активности скрывается. Файл OAuth2Service.swift разбит на подфайлы (OAuthTokenResponseBody, URLSession) для разделения ответственности.

Commit:
Создан файл  ProfileService.swift в котором реализована логика получения базовой информации о профиле

Commit:
1. Заменино API KVO для WKWebView.estimatedProgress.
2. Имплементировано в расширении URLSession дженерик-метод objectTask для уменьшения дублирования кода, связанного с обработкой результатов запроса в URLSession.dataTask().
3. Заменино использование dataTask на objectTask.
4. При получении ошибки в сетевом запросе в SplashViewController показывается UIAlertController с заголовком «Что-то пошло не так(», подписью «Не удалось войти в систему» и кнопкой «Ок» для скрытия алерта.

Commit:
Добавлен SwiftKeychainWrapper в проект через SPM.
Замено использование OAuth2TokenStorage на хранение и получение из хранилища токена с использованием SwiftKeychainWrapper.

Commit:
Промежуточный коммит -  Вёрстка кодом. Создаём View Controller

Commit:
Создан TabBarController из кода и удален TabBarController из Storyboard

Commit:
Проект запускается и не вылетает. Tab bar controller делает переход от ленты к профилю пользователя.

Commit:
исправен цвет StatusBar согласно макету и убран лишний код

mmbabaev commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -30,11 +31,56 @@ final class AuthViewController: UIViewController {
+                self.oAuthStorage.token = token
+                self.delegate?.authViewController(self, didAuthenticateWithCode: code)
+            case .failure(let error):
+                print(error)
```
нет обработки ошибки

mmbabaev commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -30,11 +31,56 @@ final class AuthViewController: UIViewController {
+
+//reserv
+
+////MARK: - protocol AuthViewControllerDelegate
```
лучше удалять закомментированный код

mmbabaev commented on ImageFeed/Auth/WebViewController.swift:
```diff
@@ -76,6 +85,15 @@ final class WebViewController: UIViewController {
+                self.progressView.isHidden = true
+            })
+        }
         progressView.setProgress(Float(webView.estimatedProgress), animated: true)
```
здесь вызывается функция с animated: true, но она вызывается много раз. Из-за этого прогресс выглядит так, что прогресс уезжает назад

mmbabaev commented on ImageFeed/Image/ImagesListService.swift:
```diff
@@ -0,0 +1,80 @@
+                switch result {
+                case .success(let photoResult):
+                    var photosArray: [Photo] = []
+                    for photo in photoResult {
```
это мелочь, но тут можно использовать функцию map:

`let photosArray = photoResult.map { Photo(photoResult: $0) }`

mmbabaev commented on ImageFeed/Image/ImagesListService.swift:
```diff
@@ -0,0 +1,80 @@
+                            object: self,
+                            userInfo: ["photos": self.photos])
+                case .failure(let error):
+                    print(error)
```
нет обработки ошибки

mmbabaev commented on ImageFeed/Image/Photo.swift:
```diff
@@ -0,0 +1,32 @@
+        self.height = CGFloat(photoResult.height)
+        self.createdAt = photoResult.createdAt
+        self.welcomeDescription = photoResult.welcomeDescription
+        self.thumbImageURL = photoResult.urls[Urls.CodingKeys.thumbImageURL.rawValue]!
```
лучше стараться избегать форс анрапов

mmbabaev commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,60 @@
+                completion(.success(()))
+
+            case .failure:
+                self.lastToken = nil
```
нет обработки ошибки (не вызван completion)

mmbabaev commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,48 @@
+                case .failure(let error):
+                    completion(.failure(error))
+                    print("error")
+                    return
```
return можно опустить

mmbabaev commented on ImageFeed/ServiceAccessLayer/OAuth2/URLSession.swift:
```diff
@@ -0,0 +1,51 @@
+                    return
+                }
+            }
+            guard let data = data else { return }
```
здесь в else лучше тоже вызвать completion

mmbabaev commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -6,62 +6,117 @@
                 switch result {
                 case .success:
-                    self?.switchToTabBarController()
+                    ProfileImageService.shared.fetchProfileImageURL(username: (self.profileService.profile?.username)!, token: token) { _ in }
```
форс анрапы

Commit:
1. 	В ImageFeed/Auth/AuthViewController.swift исправлена реализация расширения метода webViewController.
2. 	ImageFeed/Auth/WebViewController.swift исправлена реализация метода updateProgress.
3. 	В ImageFeed/Image/ImagesListService.swift изменена реализация метода fetchPhotosNextPage и добавлена обработка ошибки.
4. 	ImageFeed/Image/Photo.swift в инициализации структуры Photo убраны форс анрапы.
5.	ImageFeed/Profile/ProfileImageService.swift в методе fetchProfileImageURL исправлена реализации обработки ошибки.
6. 	ImageFeed/ServiceAccessLayer/OAuth2/URLSession.swift исправлена реализация расширения метода objectTask.
7. 	В ImageFeed/SplashView/SplashViewController.swift метод fetchProfile избавлен от форс анрапов.
8.	Исправлена ошибка grant_type при попытки авторизации, заменен адрес на https://unsplash.com/oauth/authorize
9.	У экрана Auth изменена презентация на fullscreen.

mmbabaev commented:
delegate у WebAuthController == nil
<img width="868" alt="Screenshot 2023-06-26 at 22 35 08" src="https://github.com/Gvozdkov/ImageFeed/assets/12600807/c1e6914a-8499-4f1e-bbd0-3169b2425acd">
